### ISSUE TITLE
Sqlite Database's close method is still called from a GC thread even when using ManagedDatabase

### ISSUE THREAD

**Issue by ferencdg** [bosagora/agora#1122]
There are at least 2 bugs in the current implementation of ManagedDatabase. These bugs cause unittest failures on the latest stable version of some linux distributions(Debian and Ubuntu). I will soon post code here on how to reproduce the issue and also to suggest a solution to them.

**Comment by ferencdg** [comment]
code to reproduce issue number 1

```D
module agora.utils.SimpleTest;
import std.stdio;
import std.conv;
import std.concurrency;
import agora.common.ManagedDatabase;
import core.memory : GC;
import core.thread;
import core.thread.osthread;
void simple_thread(){
    auto myDb = new ManagedDatabase(":memory:");
    writeln("thread ending");
}
void print_mem(string prefix){
    auto stat = GC.stats();
    writeln(prefix," -> used - ", stat.usedSize);
}
private void main ()
{
    print_mem("main start");
    spawn(&simple_thread);
    Thread.sleep( dur!("seconds")( 5 ) ); 
    print_mem("before GC"); // to make sure GC is actually called
    GC.collect();
    print_mem("afterGC"); // to make sure GC is actually called
    return ;
}
```

also throw an exception from the static destructor(this can happen if the Database.close() throws):
![image](https://user-images.githubusercontent.com/19815616/90467721-146b7f80-e13f-11ea-96f4-0f2e32161921.png)

this produces the following output:
![image (1)](https://user-images.githubusercontent.com/19815616/90467784-411f9700-e13f-11ea-9704-6b801119b6cb.png)

The issue here is that throwing from static destructor shouldn't be allowed, because the subsequent Database object will not be destroyed on a NON-GC thread, instead they will be destroyed on a GC thread.

**Comment by ferencdg** [comment]
The second issue is that an object that is GC allocated on threadA can outlive the destruction of threadA and garbage collected after that. This is a problem because it means ManagedDatabase can outlive thread destruction and so it will still hold a reference to the Database object(through its member variable called this.database) This means that the Database object will be destoryed on a GC thread.

```D
void simple_thread(){
    //auto s = new S(3);
    foreach(i;1..2)
        auto myDb = new ManagedDatabase(":memory:");
    writeln("thread ending");
}
void print_mem(string prefix){
    auto stat = GC.stats();
    writeln(prefix," -> used - ", stat.usedSize);
}
private void main ()
{
    print_mem("main start");
    spawn(&simple_thread);
    Thread.sleep( dur!("seconds")( 5 ) ); 
    print_mem("before GC"); // to make sure GC is actually called
    GC.collect();
    print_mem("afterGC"); // to make sure GC is actually called
    return ;
}
```

output:
![image (2)](https://user-images.githubusercontent.com/19815616/90468220-4df0ba80-e140-11ea-9226-24f2990a5f7f.png)

As you can see ManagedDatabase outlived the thread destruction. In this particular case the actual Database object was destroyed before the thread destroyed, but there is no language guarantee that this will happen. The code in the static destructor somehow detected that although there are 2 references poining to the Database objects, none of those references are rooted, so it destroyed the Database object. I don't see any language guarantee that this will always happen, and under heavy CPU/memory load it seems Database object is destroyed(in a GC thread) after thread destruction.

**Comment by ferencdg** [comment]
instead of relying on the GC, I will modify the code to use non-GC allocated heap

**Comment by Geod24** [comment]
Fixed in #1124
